core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```
@@ -140,157 +140,164 @@ public void startRecovery(final IndexShard indexShard, final DiscoveryNode sourc
         if (newTarget != null) {
             threadPool.schedule(retryAfter, ThreadPool.Names.GENERIC, new RecoveryRunner(newTarget.recoveryId()));
         }
     }
 
-    private void doRecovery(final RecoveryRef recoveryRef) {
-        RecoveryTarget recoveryTarget = recoveryRef.status();
-        assert recoveryTarget.sourceNode() != null : "can't do a recovery without a source node";
-
-        logger.trace("collecting local files for {}", recoveryTarget);
-        Store.MetadataSnapshot metadataSnapshot = null;
-        try {
-            if (recoveryTarget.indexShard().indexSettings().isOnSharedFilesystem()) {
-                // we are not going to copy any files, so don't bother listing files, potentially running
-                // into concurrency issues with the primary changing files underneath us.
-                metadataSnapshot = Store.MetadataSnapshot.EMPTY;
-            } else {
-                metadataSnapshot = recoveryTarget.indexShard().snapshotStoreMetadata();
-            }
-        } catch (org.apache.lucene.index.IndexNotFoundException e) {
-            // happens on an empty folder. no need to log
-            metadataSnapshot = Store.MetadataSnapshot.EMPTY;
-        } catch (IOException e) {
-            logger.warn("error while listing local files, recover as if there are none", e);
-            metadataSnapshot = Store.MetadataSnapshot.EMPTY;
-        } catch (Exception e) {
-            // this will be logged as warning later on...
-            logger.trace("unexpected error while listing local files, failing recovery", e);
-            onGoingRecoveries.failRecovery(recoveryTarget.recoveryId(),
-                    new RecoveryFailedException(recoveryTarget.state(), "failed to list local files", e), true);
+    private void doRecovery(final long recoveryId) {
+        RecoveryRef recoveryRef = onGoingRecoveries.getRecovery(recoveryId);
```

The RecoveryTarget is needed for one more thing:  `recoveryTarget.indexShard().prepareForIndexRecovery();` which moves the recovery stage from INIT to INDEX. I've moved this call further up and made it fail the recovery/shard if this step fails. Previously it would retry in this situation, and reset the recovery stage back to INIT, but now with double recoveries being forbidden, I think we should fail early on this.